pkg/engine/decisions.go

```diff
@@ -0,0 +1,226 @@
+package engine
+
+import (
+	"fmt"
+
+	"github.com/klothoplatform/klotho/pkg/construct"
+	"github.com/klothoplatform/klotho/pkg/engine/constraints"
+	"github.com/klothoplatform/klotho/pkg/graph"
+	knowledgebase "github.com/klothoplatform/klotho/pkg/knowledge_base"
+)
+
+type (
+	// Decision is a struct that represents a decision made by the engine
+	Decision struct {
+		Level  Level
+		Result *DecisionResult
+		Action Action
+		Cause  *Cause
+	}
+	Action string
+	Cause  struct {
+		EdgeExpansion         *graph.Edge[construct.Resource]
+		EdgeConfiguration     *graph.Edge[construct.Resource]
+		OperationalResource   construct.Resource
+		ResourceConfiguration construct.Resource
+		ConstructExpansion    construct.BaseConstruct
+		Constraint            constraints.Constraint
+	}
+	DecisionResult struct {
+		Resource construct.Resource
+		Edge     *graph.Edge[construct.Resource]
+		Config   *knowledgebase.Configuration
+	}
+	Level string
+)
+
+const (
+	// LevelError is the level of a decision when an error occurs
```

Comment:
Not really a useful godoc since this is apparent from the name. Instead, maybe describe what an "error" means and what circumstances to use that instead of one of the other levels.